CARDS-1599: Subject resources can be serialized in Text (.txt) and Markdown (.md) formats

veronikaslc opened the pull request:
### Specifications:
[CARDS-1599](https://phenotips.atlassian.net/browse/CARDS-1599)

### Testing instructions:
* Create a subject
* Add some forms with answers to it
* navigate to `localhost:8080/Subjects/FULL/SUBJECT/ID.txt` to see the text serialization
* navigate to `localhost:8080/Subjects/FULL/SUBJECT/ID.md` to see the markdown serialization
* To ensure there are no regressions, also test individual form serialization by navigating to:
`localhost:8080/Forms/FORM-ID.txt` to see the text serialization
`localhost:8080/Forms/FORM-ID.md` to see the markdown serialization

Commit:
CARDS-1599: Subject resources can be serialized in Text (.txt) and Markdown (.md) formats

sdumitriu committed:
CARDS-1599: Subject resources can be serialized in Text (.txt) and Markdown (.md) formats
Bugfix: the subjects adapter factory makes the forms adapter factory not work; reorganize the APIs to allow more types of resources to be easily serialized in the future

sashaandjic reviewed (commented):
looks good
One very minor problem: after the Smoking Cessation form, when the first 2 questions are "No", there is an "__" in the place where the formula for program is supposed to be. 

![CARDS-1599_LI](https://user-images.githubusercontent.com/23243244/151376273-75cb0882-41af-48a5-9f57-76899f180f87.jpg)

otherwise, the line is not displayed below the SC form, but before the selected program:
<img width="942" alt="image" src="https://user-images.githubusercontent.com/23243244/151377560-bbbed43e-3ec9-471f-9d82-c59e3ae7b0dc.png">

sashaandjic commented:
I bit thicker separation line would look better. Right now it is barely visible.

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/SubjectToMarkdownAdapterFactory.java:
```diff
@@ -0,0 +1,68 @@
+    @Override
+    void formatForm(final Resource resource, final StringBuilder result)
+    {
+        result.append("\n\n### ").append(resource.adaptTo(CharSequence.class)).append(MD_LINE_END).append('\n');
```
Adding `### ` in front of the form serialization which has `# Questionnaire title` as the first line means that now the form title line is `#### # Questionnaire title`, generating a `h4` and prefixing the title with an unnecessary `#` in the output.

If the intent was to decrease the heading , then prefix with `###` without a trailing space.

However, forms should look the same if printed from their own page or from the subject page, so I would just remove this fragment.

```suggestion
        result.append(resource.adaptTo(CharSequence.class)).append(MD_LINE_END).append('\n');
```

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/AbstractSubjectToStringSerializer.java:
```diff
@@ -0,0 +1,181 @@
+        .filter(value -> ValueType.OBJECT.equals(value.getValueType()))
+        .map(JsonValue::asJsonObject)
+        .filter(value -> value.containsKey("jcr:primaryType"))
+            .forEach(value -> processForm(value, result));
```
We need to add an element that would suggest a clearer separation between forms.  In text mode, it can be extra newlines. In markdown, I would suggest a double horizontal line (generated by `---\n`) which we can then style from css to produce a page-break (`"& hr+hr" : {height: 0, pageBreakAfter: "always"}`). Note that we should not add a double `hr` after the last form.

marta- replied:
Actually disregard this comment. We can page-break before `h1:not(first-child)`. Instead, let's not export any subject (id, parent, date) in markdown.

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/AbstractSubjectToStringSerializer.java:
```diff
@@ -0,0 +1,181 @@
+        if (parent != null) {
+            formatMetadata(getSubjectFullIdentifier(subjectJson), result);
+        }
+        formatMetadata("\n", result);
```
Semantically, `"\n"` is not metadata. Any new lines that are needed should be introduced by the functions that format different pieces of the output. All `formatMetadata("\n", result)` should be removed. If needed, we could introduce a `formatMetadataSeparator(result)` which produces some sort of visual separator between parts of the output. However, it's not needed on this particular line.

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/AbstractSubjectToStringSerializer.java:
```diff
@@ -0,0 +1,181 @@
+            formatMetadata(getSubjectFullIdentifier(subjectJson), result);
+        }
+        formatMetadata("\n", result);
+        formatType(getSubjectType(subjectJson), result);
+        formatMetadata(getSubjectIdentifier(subjectJson), result);
```
Merge these into an abstract `formatSubjectTitle(type, identifier, result)`

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/AbstractSubjectToStringSerializer.java:
```diff
@@ -0,0 +1,181 @@
+        formatType(getSubjectType(subjectJson), result);
+        formatMetadata(getSubjectIdentifier(subjectJson), result);
+        formatMetadata("\n", result);
+        formatMetadata(getCreationDate(subjectJson), result);
```
Add a `formatCreationDate(date, result)`.

Having specific functions for different pieces of information helps hide those pieces of information in one of the serialization formats. For example, if we don't want the creation date in markdown, the implementation of `formatCreationDate` can be `{ /* Do nothing */}` without affecting how other "metadata" items are serialized.

marta- commented:
> looks good One very minor problem: after the Smoking Cessation form, when the first 2 questions are "No", there is an "__" in the place where the formula for program is supposed to be.

That's caused by https://phenotips.atlassian.net/browse/CARDS-1597, out of scope.

> otherwise, the line is not displayed below the SC form, but before the selected program: 

That's caused by the fact that it's a question without `text`, only `value`. Smoking cessation is really messy imo, and tailoring the serialization functionality to deal with it is not worth the effort. One workaround is to replace that question with 4 different info cards, which will end up only being displayed in edit mode. Please make an [IDEA] task for this with all the details (including that it looks bad in markdown serialization) and we'll see if we ever get to it.

marta- commented:
> > looks good One very minor problem: after the Smoking Cessation form, when the first 2 questions are "No", there is an "__" in the place where the formula for program is supposed to be.
> 
> That's caused by https://phenotips.atlassian.net/browse/CARDS-1597, out of scope.
> 
> > otherwise, the line is not displayed below the SC form, but before the selected program:
> 
> That's caused by the fact that it's a question without `text`, only `value`. Smoking cessation is really messy imo, and tailoring the serialization functionality to deal with it is not worth the effort. One workaround is to replace that question with 4 different info cards, which will end up only being displayed in edit mode. Please make an [IDEA] task for this with all the details (including that it looks bad in markdown serialization) and we'll see if we ever get to it.

Hm, actually we could easily just not call `formatQuestion` if `StringUtils.isBlank(question)` @veronikaslc .

veronikaslc committed:
CARDS-1599: Subject resources can be serialized in Text (.txt) and Markdown (.md) formats

Removed unneded extra styling

Co-authored-by: Marta Girdea <marta.girdea@gmail.com>

Commit:
CARDS-1599: Subject resources can be serialized in Text (.txt) and Markdown (.md) formats
* Not export any subject (id, parent, date) in markdown
* Minor code refactorings

Commit:
CARDS-1598: Form resources can be serialized in Markdown format (.md)
Do not display answer if question text is blank

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/SubjectToTextAdapterFactory.java:
```diff
@@ -46,15 +46,21 @@ public String serialize(Resource resource)
+    void formatSubjectTitle(String type, String identifier, StringBuilder result)
     {
-        result.append(metadata.toUpperCase(Locale.ROOT)).append('\n');
+        result.append(type).append("  ").append(identifier).append("\n");
```
Why 2 spaces?

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/SubjectToMarkdownAdapterFactory.java:
```diff
@@ -0,0 +1,72 @@
+    @Override
+    void formatParent(String metadata, StringBuilder result)
+    {
+        // Don't output the subject
```
```suggestion
        // Don't output the parent subject
```

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/serialize/AbstractFormToStringSerializer.java:
```diff
@@ -189,10 +189,14 @@ private void processAnswer(final JsonObject answerJson, final String nodeType, f
         }
+        final String questionText = answerJson.getJsonObject("question").getString("text");
+        if (StringUtils.isBlank(questionText)) {
+            return;
```
No, we should still display the answer if there is one.

Commit:
CARDS-1598: Form resources can be serialized in Markdown format (.md)
Do not display question as empty if question text is blank

Commit:
CARDS-1599: Subject resources can be serialized in Text (.txt) and Markdown (.md) formats
* Minir addressing comments

marta- committed:
CARDS-1598: Form resources can be serialized in Markdown format (.md)
Do not display an empty answer if the question text is blank

marta- committed:
CARDS-1598: Form resources can be serialized in Markdown format (.md)
Reorganize how newlines are generated for a tidier output

marta- committed:
CARDS-1599: Subject resources can be serialized in Text (.txt) and Markdown (.md) formats
Reordered functions for consistency with the text serializer

marta- committed:
CARDS-1599: Subject resources can be serialized in Text (.txt) and Markdown (.md) formats
Reorganize how newlines are generated for a tidier output
Simplify subject metadata in text serialization
Minor code style improvements (parameter names, order of functions)

sashaandjic commented:
One very small thing: if subject does not have any forms, the txt has the subject header, while md is empty.

![empty-visit-md](https://user-images.githubusercontent.com/23243244/151647795-21a52c97-6e07-49ea-987d-e5bc314c8a8c.png)
![empty-visit-txt](https://user-images.githubusercontent.com/23243244/151647799-9aa63480-1e9e-43b8-9794-526b699892e2.png)
-------------
Visit with long forms
![long-forms-md](https://user-images.githubusercontent.com/23243244/151647816-a80ae7c1-b7cb-4211-9e4d-f39bbc18acb4.png)
[long-forms-txt.txt](https://github.com/data-team-uhn/cards/files/7962992/long-forms-txt.txt)

marta- commented:
> **Problem**: txt starts with subject and date header, while md does not.

This is intentional. Printing out only the forms without any subject information will simply look better. I'll be trying to add the subject information when generating the print preview, but it will not be from markdown serialization so that it can be styled better.

> **Question**: Can we add PatientID to the Visit header?

I just removed it in the last commit because it didn't look nice to me. I can add it back, but it's really low priority.

sashaandjic reviewed (approved):
tested
